tests/acceptance/chains-test.js

```diff
@@ -0,0 +1,342 @@
+		this.service = this.appInstance.lookup('service:promise-chain');
+
+		this.appInstance.register('route:A', Route.extend());
+		this.appInstance.register('route:B', Route.extend());
+		this.appInstance.register('route:A.A', Route.extend());
+		this.appInstance.register('route:A.B', Route.extend());
+		this.appInstance.register('route:B.A', Route.extend());
+		this.appInstance.register('route:A.A.A', Route.extend());
+
+		injectPromiseChain(this.appInstance);
+
+		return visit('/');
+	}
+});
+
+// Flow.
+
+test('it runs "onEnter" hook upon entering a route', async function(assert) {
+	const routeA = this.appInstance.lookup('route:A');
+
+	routeA.onEnter = this.sandbox.spy();
+
+	await visit('/A');
+
+	assert.ok(routeA.onEnter.calledOnce, 'hook is called');
+});
+
+test('it runs "onExit" hook upon exiting a route', async function(assert) {
+	const routeA = this.appInstance.lookup('route:A');
+
+	await visit('/A');
+
+	routeA.onExit = this.sandbox.spy();
+
+	await visit('/B');
+
+	assert.ok(routeA.onExit.calledOnce, 'hook is called');
+});
+
+test('it runs "onEnter" hook on every nested routes', async function(assert) {
+	const routeA = this.appInstance.lookup('route:A');
+	const routeAA = this.appInstance.lookup('route:A.A');
+
+	routeA.onEnter = this.sandbox.spy();
+	routeAA.onEnter = this.sandbox.spy();
+
+	await visit('/A/A');
+
+	assert.ok(routeA.onEnter.calledOnce, 'hook is called');
+	assert.ok(routeAA.onEnter.calledOnce, 'hook is called');
+});
+
+test('it runs "onExit" hook from nested routes', async function(assert) {
+	const routeA = this.appInstance.lookup('route:A');
+	const routeAA = this.appInstance.lookup('route:A.A');
+
+	await visit('/A/A');
+
+	routeA.onExit = this.sandbox.spy();
+	routeAA.onExit = this.sandbox.spy();
+
+	await visit('/B');
+
+	assert.ok(routeA.onExit.calledOnce, 'hook is called');
+	assert.ok(routeAA.onExit.calledOnce, 'hook is called');
+});
+
+test('it runs "onEnter" hook only on changed routes', async function(assert) {
+	const routeA = this.appInstance.lookup('route:A');
+	const routeAB = this.appInstance.lookup('route:A.B');
+
+	await visit('/A/A');
+
+	routeA.onEnter = this.sandbox.spy();
+	routeAB.onEnter = this.sandbox.spy();
+
+	await visit('/A/B');
+
+	assert.ok(routeA.onEnter.notCalled, 'hook is not called');
+	assert.ok(routeAB.onEnter.calledOnce, 'hook is called');
+});
+
+test('it runs "onExit" hook only on changed routes', async function(assert) {
+	const routeA = this.appInstance.lookup('route:A');
+	const routeAA = this.appInstance.lookup('route:A.A');
+
+	await visit('/A/A');
+
+	routeA.onExit = this.sandbox.spy();
+	routeAA.onExit = this.sandbox.spy();
+
+	await visit('/A/B');
+
+	assert.ok(routeA.onExit.notCalled, 'hook is not called');
+	assert.ok(routeAA.onExit.calledOnce, 'hook is called');
+});
+
+test('it does not call unnecesary hooks', async function(assert) {
+	const routeB = this.appInstance.lookup('route:B');
+	const routeAAA = this.appInstance.lookup('route:A.A.A');
```

Comment:
It's like batteries O_O
<img src="https://images-na.ssl-images-amazon.com/images/I/41NfAdxwO%2BL.jpg" width="200">